rskj-core/src/main/java/co/rsk/peg/PendingFederation.java

```diff
@@ -0,0 +1,144 @@
+ * voted by the current federation to potentially become
+ * the new active federation.
+ *
+ * @author Ariel Mendelzon
+ */
+public final class PendingFederation {
+    private long id;
+    private int numberOfSignaturesRequired;
+    private List<BtcECKey> publicKeys;
+
+    public PendingFederation(long id, int numberOfSignaturesRequired, List<BtcECKey> publicKeys) {
+        this.id = id;
+        this.numberOfSignaturesRequired = numberOfSignaturesRequired;
+        // Sorting public keys ensures same order of federators for same public keys
+        // Immutability provides protection unless unwanted modification, thus making the Pending Federation instance
+        // effectively immutable
+        this.publicKeys = Collections.unmodifiableList(publicKeys.stream().sorted(BtcECKey.PUBKEY_COMPARATOR).collect(Collectors.toList()));
+    }
+
+    public long getId() {
+        return id;
+    }
+
+    public List<BtcECKey> getPublicKeys() {
+        return publicKeys;
+    }
+
+    public int getNumberOfSignaturesRequired() {
+        return this.numberOfSignaturesRequired;
+    }
+
+    public boolean isComplete() {
+        return this.publicKeys.size() >= this.numberOfSignaturesRequired;
+    }
+
+    /**
+     * Creates a new PendingFederation with the additional specified public key
+     * @param key the new public key
+     * @return a new PendingFederation with the added public key
+     */
+    public PendingFederation addPublicKey(BtcECKey key) {
+        List<BtcECKey> newKeys = new ArrayList<>(publicKeys);
+        newKeys.add(key);
+        return new PendingFederation(this.id, this.numberOfSignaturesRequired, newKeys);
+    }
+
+    /**
+     * Creates a new PendingFederation without the specified public key
+     * @param key the public key to remove
+     * @return a new PendingFederation without the given public key
+     */
+    public PendingFederation removePublicKey(BtcECKey key) {
+        if (!publicKeys.contains(key)) {
+            throw new IllegalStateException("PendingFederation doesn't contain the given public key");
+        }
+
+        List<BtcECKey> newKeys = new ArrayList<>(publicKeys);
+        newKeys.remove(key);
+        return new PendingFederation(this.id, this.numberOfSignaturesRequired, newKeys);
+    }
+
+    public Federation buildFederation(Instant creationTime, NetworkParameters btcParams) {
+        if (!this.isComplete()) {
+            throw new IllegalStateException("PendingFederation is incomplete");
+        }
+
+        return new Federation(
+                numberOfSignaturesRequired,
+                publicKeys,
+                creationTime,
+                btcParams
+        );
+    }
+
+    @Override
+    public String toString() {
+        return String.format("%d of %d signatures pending federation (%s)", numberOfSignaturesRequired, publicKeys.size(), isComplete() ? "complete" : "incomplete");
+    }
+
+    @Override
+    public boolean equals(Object other) {
```

Comment:
Overriding only equals violates the [hashcode](http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html#hashCode%28%29) contract of the Object. As a general rule, both methods (equals and hashcode) must be overridden together

Reply:
Where's this being used? Can't we remove the @Override?

Reply:
Not being used. What's the normal way to implement `hashCode`, @diega ?